SU2_CFD/include/numerics_simd/CNumericsSIMD.hpp

```diff
@@ -0,0 +1,711 @@
+    const auto jPoint = geometry.edges->GetNode<Int>(iEdge,1);
+    const auto iEdges = Int(iEdge,1);
+
+    /*--- Geometric properties. ---*/
+
+    const auto vector_ij = distanceVector<nDim>(iPoint, jPoint, geometry.nodes->GetCoord());
+
+    const auto normal = gatherVariables<nDim>(iEdges, geometry.edges->GetNormal());
+    const auto area = norm(normal);
+    VectorDbl<nDim> unitNormal;
+    for (size_t iDim = 0; iDim < nDim; ++iDim) {
+      unitNormal(iDim) = normal(iDim) / area;
+    }
+
+    /*--- Reconstructed primitives. ---*/
+
+    auto V = getCompressiblePrimitives(iPoint, jPoint, config.GetMUSCL_Flow(),
+                                       static_cast<ENUM_LIMITER>(config.GetKind_SlopeLimit_Flow()),
+                                       vector_ij, solution);
+
+    /*--- Compute conservative variables. ---*/
+
+    CPair<CCompressibleConservatives<nDim> > U;
+    U.i = getCompressibleConservatives(V.i);
+    U.j = getCompressibleConservatives(V.j);
+
+    /*--- Roe averaged variables. ---*/
+
+    auto roeAvg = getRoeAveragedVariables(gamma, V, unitNormal);
+
+    /*--- P tensor. ---*/
+
+    auto pMat = pMatrix(gamma, roeAvg.density, roeAvg.velocity,
+                        roeAvg.projVel, roeAvg.speedSound, unitNormal);
+
+    /*--- Grid motion. ---*/
+
+    Double projGridVel = 0.0, projVel = roeAvg.projVel;
+    if (dynamicGrid) {
+      const auto& gridVel = geometry.nodes->GetGridVel();
+      projGridVel = 0.5*(dot(gridVel.innerIter<nDim>(iPoint), unitNormal)+
+                         dot(gridVel.innerIter<nDim>(jPoint), unitNormal));
+      projVel -= projGridVel;
+    }
+
+    /*--- Convective eigenvalues. ---*/
+
+    VectorDbl<nVar> lambda;
+    for (size_t iDim = 0; iDim < nDim; ++iDim) {
+      lambda(iDim) = projVel;
+    }
+    lambda(nDim) = projVel + roeAvg.speedSound;
+    lambda(nDim+1) = projVel - roeAvg.speedSound;
+
+    /*--- Apply Mavriplis' entropy correction to eigenvalues. ---*/
+
+    auto maxLambda = abs(projVel) + roeAvg.speedSound;
+
+    for (size_t iVar = 0; iVar < nVar; ++iVar) {
+      lambda(iVar) = simd::max(abs(lambda(iVar)), entropyFix*maxLambda);
+    }
+
+    /*--- Inviscid fluxes and Jacobians. ---*/
+
+    auto flux_i = inviscidProjFlux(V.i, U.i, unitNormal);
+    auto flux_j = inviscidProjFlux(V.j, U.j, unitNormal);
+
+    VectorDbl<nVar> flux;
+    for (size_t iVar = 0; iVar < nVar; ++iVar) {
+      flux(iVar) = kappa * (flux_i(iVar) + flux_j(iVar));
+    }
+
+    MatrixDbl<nVar> jac_i, jac_j;
+    if (implicit) {
+      jac_i = inviscidProjJac(gamma, V.i.velocity(), U.i.rhoEnergy()/U.i.density(), normal, kappa);
+      jac_j = inviscidProjJac(gamma, V.j.velocity(), U.j.rhoEnergy()/U.j.density(), normal, kappa);
+    }
+
+    /*--- Finalize in derived class (static polymorphism). ---*/
+
+    Derived::finalizeResidual(flux, jac_i, jac_j, implicit, gamma, kappa, area,
+                              unitNormal, V, U, roeAvg, lambda, pMat);
```

Comment:
Then we can call static methods of those derived classes like this.
Being static means we need to pass all arguments explicitly, but as mentioned above we do not want to rely on member variables for "Base-Derived communication" as that would not be thread-safe.